src/libopensc/card-piv.c

```diff
@@ -4378,6 +4384,75 @@ static int piv_get_pin_preference(sc_card_t *card, int *pin_ref)
 	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
 }
 
+static void
+piv_yk_metadata_get_policy(sc_context_t *ctx, u8 *buf, size_t buflen, u8 *pin, u8 *touch)
+{
+	size_t policylen;
+	const u8 *policy = sc_asn1_find_tag(ctx, buf, buflen, 0x02, &policylen);
+	if (policy && policylen == 2) {
+		*pin = policy[0];
+		*touch = policy[1];
+	}
+}
+
+static void
+piv_yk_get_metadata(sc_card_t *card, u8 slot, u8 *pin_policy, u8 *touch_policy)
+{
+	sc_apdu_t apdu;
+	u8 resp[100];
+	size_t i;
+	piv_private_data_t *priv = PIV_DATA(card);
+
+	/* initialize with the default behaviour */
+	if (pin_policy)
+		*pin_policy = 0x00;
+	if (touch_policy)
+		*pin_policy = 0x00;
+
+	if (priv->yubico_version < 0x00050300) {
+		sc_log(card->ctx, "Yubikey's PIN and touch policy not available");
+		return;
+	}
+
+	for (i = 0; i < (sizeof(priv->yk_metadata)/sizeof(*priv->yk_metadata)); i++) {
+		if (priv->yk_metadata[i].slot == 0x00)
+			/* reached the last initialized entry */
+			break;
+
+		if (priv->yk_metadata[i].slot == slot)
+			/* metadata already initialized */
+			break;
+	}
+
+	if (priv->yk_metadata[i].slot == 0x00) {
```

Comment:
if we fill the whole 26 slots in the metedata array, you will access memory behind the buffer. Even though well behaving yubikey should not have more than 26 slots (I think), it would be good to add a sanity check the `i` is smaller than the length of the buffer here.

Reply:
the plan was to stop when `i` is exactly the last index. i changed the loop to fix this, thanks